platform/o.n.bootstrap/src/org/netbeans/JarClassLoader.java

```diff
@@ -574,6 +606,17 @@ protected URL doGetResource(String name) throws IOException  {
         @Override
         protected byte[] readClass(String path) throws IOException {
             try {
+                if (isMultiRelease() && RUNTIME_VERSION != BASE_VERSION)
+                {
+                    int ver = RUNTIME_VERSION;
+                    while (ver > BASE_VERSION) {
+                        byte[] data = archive.getData(this, "META-INF/versions/" + ver + "/" + path);
+                        if (data != null) {
+                            return data;
+                        }
+                        ver--;
+                    }
+                }
```

Comment:
This might scale badly. Lets say we load lucene which has a 8 and a 9 version and run on JDK 19.

This will call `getData()` (which isn't trivial) 10 times for each class before it finds something. I would put all versions contained in `META-INF/versions/` into an array during init and go through the array backwards. This would produce hits with fewer attempts on average.

Reply:
looking at the JDK impl

https://github.com/openjdk/jdk/blob/0bcf17674e1c8c07ede3aef6449e0631e526ba86/src/java.base/share/classes/java/util/jar/JarFile.java#L599-L623

JarFile lists all versions first and iterates only through them, which doesn't scale negatively with the runtime version.

Reply:
Good point. Fixed.